File: C++/count-primes.cpp
// Time:  O(n)
// Space: O(n)

class Solution {
public:
    int countPrimes(int n) {
        return size(linear_sieve_of_eratosthenes(n - 1));
    }

private:
    vector<int64_t> linear_sieve_of_eratosthenes(int64_t n) {  // Time: O(n), Space: O(n)
        vector<int64_t> spf(n + 1, -1);
        vector<int64_t> primes;
        for (int64_t i = 2; i <= n; ++i) {
            if (spf[i] == -1) {
                spf[i] = i;
                primes.emplace_back(i);
            }
            for (const auto& p : primes) {
                if (i * p > n || p > spf[i]) {
                    break;
                }
                spf[i * p] = p;
            }
        }
        return primes;
    }
};

// Time:  O(n/2 + n/3 + ... + n/p) = O(nlog(logn)), see https://mathoverflow.net/questions/4596/on-the-series-1-2-1-3-1-5-1-7-1-11
// Space: O(n)
class Solution2 {
public:
    int countPrimes(int n) {
        if (n <= 2) {
            return 0;
        }

        vector<int> is_prime(n / 2, true);
        auto cnt = is_prime.size();
        for (int i = 3; i * i < n; i += 2) {
            if (!is_prime[i / 2]) {
                continue;
            }
            for (int j = i * i; j < n; j += 2 * i) {
                if (!is_prime[j / 2]) {
                    continue;
                }
                --cnt;
                is_prime[j / 2] = false;
            }
        }
        return cnt;
    }
};

File: Python/count-primes.py
# Time:  O(n/2 + n/3 + ... + n/p) = O(nlog(logn)), see https://mathoverflow.net/questions/4596/on-the-series-1-2-1-3-1-5-1-7-1-11
# Space: O(n)

class Solution(object):
    # @param {integer} n
    # @return {integer}
    def countPrimes(self, n):
        if n <= 2:
            return 0

        is_prime = [True]*(n//2)
        cnt = len(is_prime)
        for i in xrange(3, n, 2):
            if i * i >= n:
                break
            if not is_prime[i//2]:
                continue
            for j in xrange(i*i, n, 2*i):
                if not is_prime[j//2]:
                    continue
                cnt -= 1
                is_prime[j//2] = False

        return cnt


# Time:  O(n)
# Space: O(n)
class Solution_TLE(object):
    def countPrimes(self, n):
        """
        :type n: int
        :rtype: int
        """
        def linear_sieve_of_eratosthenes(n):
            primes = []
            spf = [-1]*(n+1)  # the smallest prime factor
            for i in xrange(2, n+1):
                if spf[i] == -1:
                    spf[i] = i
                    primes.append(i)
                for p in primes:
                    if i*p > n or p > spf[i]:
                        break
                    spf[i*p] = p
            return primes
    
        return len(linear_sieve_of_eratosthenes(n-1))
